server/py/services/api/initial_data.py

```diff
@@ -877,41 +877,104 @@ def _perform_version_8_data_migrations(
-        .filter(framework.db.sqldb.models.Function.kind.is_(None))
-        .count()
+    def handle_function_kind(record):
+        function_dict = record.struct
+        record.kind = function_dict.pop("kind", None)
+        if record.kind is not None:
+            record.struct = function_dict
+            return record
+        return None
+
+    def filter_function_kind():
+        return getattr(framework.db.sqldb.models.Function, "kind").is_(None)
+
+    _migrate_data(
+        db,
+        db_session,
+        model=framework.db.sqldb.models.Function,
+        filter_func=filter_function_kind,
+        handle_field_record_func=handle_function_kind,
+        chunk_size=chunk_size,
     )
-    logger.info(f"Number of functions to migrate: {number_of_functions_to_migrate}")
-    while True:
-        logger.info("Fetching functions to migrate", offset=offset)
-        q = db._query(db_session, framework.db.sqldb.models.Function)
-        q = q.order_by(framework.db.sqldb.models.Function.id)
-        q = q.limit(chunk_size)
-        q = q.offset(offset)
-        functions = []
-        for function in q.all():
-            function_dict = function.struct
-            function.kind = function_dict.pop("kind", None)
-            if function.kind is None:
-                continue
-            function.struct = function_dict
-            functions.append(function)
-        if functions:
-            logger.info("Committing migrated functions", count=len(functions))
-            db_session.add_all(functions)
-            db._commit(db_session, functions)
-        if len(functions) < chunk_size:
-            logger.info("No more functions to migrate")
+
+
+def _migrate_artifact_producer_uri(
+    db: framework.db.sqldb.db.SQLDB,
+    db_session: sqlalchemy.orm.Session,
+    chunk_size: int = 500,
+):
+    def handle_artifact_producer_uri(record):
+        record.producer_uri = (
+            record.full_object.get("spec", {}).get("producer", {}).get("uri", None)
+        )
+        if record.producer_uri is not None:
+            return record
+        return None
+
+    def filter_artifacts():
+        return getattr(framework.db.sqldb.models.ArtifactV2, "producer_uri").is_(None)
+
+    _migrate_data(
+        db,
+        db_session,
+        model=framework.db.sqldb.models.ArtifactV2,
+        filter_func=filter_artifacts,
+        handle_field_record_func=handle_artifact_producer_uri,
+        chunk_size=chunk_size,
+    )
+
+
+def _migrate_data(
+    db: framework.db.sqldb.db.SQLDB,
+    db_session: sqlalchemy.orm.Session,
+    model,
+    filter_func,
+    handle_field_record_func,
+    chunk_size: int = 500,
+):
+    # Query for records that need migration
+    records = db._query(db_session, model).filter(filter_func).limit(chunk_size).all()
+
+    if not records:
+        logger.info(f"No records to migrate for {model.__name__.lower()}")
+        return
+
+    logger.info(f"Starting migration for {len(records)} {model.__name__.lower()}s")
```

Comment:
since `artifactv2s` looks weird, I'd go with:
```suggestion
    logger.info(f"Starting migration for {len(records)} {model.__name__.lower()} records")
```